nixos/modules/security/wrappers/wrapper.c

```diff
@@ -161,22 +162,28 @@ static int make_caps_ambient(const char *selfPath)
 
 int main(int argc, char * * argv)
 {
-    // I *think* it's safe to assume that a path from a symbolic link
-    // should safely fit within the PATH_MAX system limit. Though I'm
-    // not positive it's safe...
-    char selfPath[PATH_MAX];
-    int selfPathSize = readlink("/proc/self/exe", selfPath, sizeof(selfPath));
+    // O_PATH | O_NOFOLLOW gives us a fd pointing to the symlink
+    int selfExe = open("/proc/self/exe", O_PATH | O_CLOEXEC | O_NOFOLLOW);
+    assert(selfExe != -1);
+    struct stat st;
+    assert(fstat(selfExe, &st) != -1);
+    size_t selfPathCap = st.st_size + 1;
+    char *selfPath = malloc(selfPathCap);
+    assert(selfPath);
+    int selfPathSize = readlinkat(selfExe, "", selfPath, selfPathCap);
 
     assert(selfPathSize > 0);
 
     // Assert we have room for the zero byte, this ensures the path
     // isn't being truncated because it's too big for the buffer.
     //
-    // A better way to handle this might be to use something like the
-    // whereami library (https://github.com/gpakosz/whereami) or a
-    // loop that resizes the buffer and re-reads the link if the
-    // contents are being truncated.
-    assert(selfPathSize < sizeof(selfPath));
+    // I don't think it's possible for the link contents to change
+    // between opening a symlink fd and readlinkat on it, so this is
+    // probably not necessary. Doubly so since this is /proc/self/exe,
+    // not a normal symlink. But it's trivial to check.
+    assert(selfPathSize < selfPathCap);
```

Comment:
Shouldn't this always fail? I mean, `selfPathSize` is the number of bytes put into the buffer, terminating-zero included, and `selfPathCap` is the size _without_ terminating zero, plus 1.

Reply:
No, `selfPathSize` is the result of `readlinkat` which also lacks the `0`.

Reply:
I'll update the comments with this though.

Reply:
Just in case there were not enough ways to start a heap walk… Missed that, yes, my bad.